### PR TITLE
Splitting the backup tests in create/restore/delete

### PR DESCRIPTION
**Pull request by izaac**
This is a refactor of https://github.com/rancher/rancher/pull/24781 which now will split on create, restore and delete tests.

Functionality wasn't changed and tests were executed in Jenkins for local filesystem tests and s3 with credentials to be sure it's working.

### CODE REVIEW

**Comment by izaac** [comment]
In the latest change I'm using the `pytest ordering` plugin to ensure order instead of test naming, sample using just `print` statements to validate the ordering:

```bash
tests/v3_api/test_bkp_restore.py::test_bkp_restore_local_create PASSED                                                                                                 [  8%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_local_restore PASSED                                                                                                [ 16%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_local_delete PASSED                                                                                                 [ 25%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_local_with_snapshot_check_create PASSED                                                                             [ 33%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_local_with_snapshot_check_restore PASSED                                                                            [ 41%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_local_with_snapshot_check_delete PASSED                                                                             [ 50%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_s3_with_creds_create PASSED                                                                                         [ 58%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_s3_with_creds_restore PASSED                                                                                        [ 66%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_s3_with_creds_delete PASSED                                                                                         [ 75%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_s3_with_iam_create PASSED                                                                                           [ 83%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_s3_with_iam_restore PASSED                                                                                          [ 91%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_s3_with_iam_delete PASSED                                                                                           [100%]
```

We still need this https://github.com/rancher/rancher/pull/24984 fix before running the one click testing in Jenkins.

**Comment by izaac** [comment]
No more dependencies. Ready for review.

File: tests/validation/tests/v3_api/test_bkp_restore.py [hunk 1]
```diff
@@ -8,79 +8,142 @@
...
+def test_bkp_restore_local_create(create_project_client):
+    validate_backup_create("local")
 
-def test_bkp_restore_local(create_project_client):
-    validate_backup_create_restore_delete("local")
+
+@pytest.mark.run(order=2)
+def test_bkp_restore_local_restore():
+    validate_backup_restore("local")
+
```
Comment (sangeethah) [inline tests/validation/tests/v3_api/test_bkp_restore.py]: @izaac We should still have the ability to run each of the cases  relating to restore and delete individually if we choose to. Can you have these methods also include all the pre req steps needed like create for these test cases ?

Comment (izaac) [reply tests/validation/tests/v3_api/test_bkp_restore.py]: Ok I would have to re-design the splitting.  Will work on that.

File: tests/validation/tests/v3_api/test_bkp_restore.py [hunk 2]
```diff
@@ -8,79 +8,142 @@
 
 namespace = {"p_client": None, "ns": None, "cluster": None, "project": None,
              "nodes": []}
+backup_info = {"backupname": None, "backup_id": None, "workload": None,
+               "backupfilename": None, "etcdbackupdata": None}
 
+@pytest.mark.run(order=1)
+def test_bkp_restore_local_create(create_project_client):
+    validate_backup_create("local")
```
Comment (sowmyav27) [inline tests/validation/tests/v3_api/test_bkp_restore.py]: Are we handling "local" case in create/delete case?

File: tests/validation/tests/v3_api/test_bkp_restore.py [hunk 3]
```diff
@@ -90,37 +153,41 @@ def validate_backup_create_restore_delete(backup_mode):
     validate_workload(p_client, testworkload, "deployment", ns.name)
 
     # Perform Restore
-    cluster.restoreFromEtcdBackup(etcdBackupId=backupId)
+    cluster.restoreFromEtcdBackup(etcdBackupId=backup_info["backup_id"])
     # After restore, validate cluster
     validate_cluster(client, cluster, intermediate_state="updating",
```
[the remark this answers is not in the document]
Comment (izaac) [reply tests/validation/tests/v3_api/test_bkp_restore.py]: I do have the cluster object already at that point, here I'm just trying make sure the cluster is in Active state again before proceeding for the next step

**Comment by izaac** [comment]
Local run output:

```bash
$ pipenv run pytest -v -k "test_bkp_restore" tests/v3_api/ -W ignore::DeprecationWarning --debug
```

```bash
writing pytestdebug information to /Users/izaac/repos/rancher-izaac/tests/validation/pytestdebug.log
========================================================================================== test session starts ==========================================================================================
platform darwin -- Python 3.7.6, pytest-5.3.2, py-1.8.1, pluggy-0.13.1 -- /Users/izaac/.local/share/virtualenvs/validation-pB2uESK5/bin/python3.7
using: pytest-5.3.2 pylib-1.8.1
setuptools registered plugins:
  pytest-xdist-1.22.0 at /Users/izaac/.local/share/virtualenvs/validation-pB2uESK5/lib/python3.7/site-packages/xdist/plugin.py
  pytest-xdist-1.22.0 at /Users/izaac/.local/share/virtualenvs/validation-pB2uESK5/lib/python3.7/site-packages/xdist/looponfail.py
  pytest-forked-1.1.3 at /Users/izaac/.local/share/virtualenvs/validation-pB2uESK5/lib/python3.7/site-packages/pytest_forked/__init__.py
cachedir: .pytest_cache
rootdir: /Users/izaac/repos/rancher-izaac/tests/validation
plugins: xdist-1.22.0, forked-1.1.3
collected 337 items / 325 deselected / 12 selected

tests/v3_api/test_bkp_restore.py::test_bkp_restore_local_create PASSED                                                                                                                            [  8%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_local_restore PASSED                                                                                                                           [ 16%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_local_delete PASSED                                                                                                                            [ 25%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_local_with_snapshot_check_create PASSED                                                                                                        [ 33%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_local_with_snapshot_check_restore PASSED                                                                                                       [ 41%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_local_with_snapshot_check_delete PASSED                                                                                                        [ 50%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_s3_with_creds_create PASSED                                                                                                                    [ 58%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_s3_with_creds_restore PASSED                                                                                                                   [ 66%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_s3_with_creds_delete PASSED                                                                                                                    [ 75%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_s3_with_iam_create PASSED                                                                                                                      [ 83%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_s3_with_iam_restore PASSED                                                                                                                     [ 91%]
tests/v3_api/test_bkp_restore.py::test_bkp_restore_s3_with_iam_delete PASSED                                                                                                                      [100%]
-- Docs: https://docs.pytest.org/en/latest/warnings.html
==================================================================== 12 passed, 325 deselected, 5723 warnings in 4029.59s (1:07:09) =====================================================================
```

File: tests/validation/tests/v3_api/test_bkp_restore.py [hunk 4]
```diff
@@ -90,37 +166,51 @@ def validate_backup_create_restore_delete(backup_mode):
...
     workload_list = p_client.list_workload(uuid=testworkload.uuid).data
     print(len(workload_list))
-    assert len(workload_list) == 0
+    assert len(workload_list) == 0, "workload shouldn't exist after restore"
 
+
+def validate_backup_delete(backup_mode):
+    client = get_user_client()
+    cluster = namespace["cluster"]
     if backup_mode == "s3":
```
[the remark this answers is not in the document]
Comment (izaac) [reply tests/validation/tests/v3_api/test_bkp_restore.py]: It doesn't do any type of file checks, neither in the local filesystem or S3 bucket. It's only for the simple scenario that triggers a backup from Rancher and waits for active.

File: tests/validation/tests/v3_api/test_bkp_restore_local.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+               "backupfilename": None, "etcdbackupdata": None}
+
+
+def test_bkp_restore_local_create():
+    validate_backup_create("local", namespace, backup_info)
+
+
+def test_bkp_restore_local_restore():
+    ns , binfo = validate_backup_create("local", namespace, backup_info)
+    validate_backup_restore("local", ns, binfo)
```
[the remark this answers is not in the document]
Comment (izaac) [reply tests/validation/tests/v3_api/test_bkp_restore_local.py]: We can pass whatever we want, "local" here means don't do any `s3`, `filesystem` checks. But I think I will remove the parameter from the function as previously suggested.

File: tests/validation/tests/v3_api/common.py [hunk 6]
```diff
@@ -1828,3 +1829,158 @@ def generate_template_global_role(name, new_user_default=False, template=None):
...
+    # after restore
+    workload_list = p_client.list_workload(uuid=testworkload.uuid).data
+    print(len(workload_list))
+    assert len(workload_list) == 0, "workload shouldn't exist after restore"
+    return namespace, backup_info
+
+
+def validate_backup_delete(backup_mode, namespace, backup_info):
+    client = get_user_client()
+    cluster = namespace["cluster"]
```
Comment (sowmyav27) [inline tests/validation/tests/v3_api/common.py]: for the case of "local" we can validate if on a delete of a backup, it is not listed in the API anymore.

File: tests/validation/tests/v3_api/test_bkp_restore_local.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+from .common import *  # NOQA
+
+namespace = {"p_client": None, "ns": None, "cluster": None, "project": None,
+             "nodes": []}
+backup_info = {"backupname": None, "backup_id": None, "workload": None,
+               "backupfilename": None, "etcdbackupdata": None}
+
+
+def test_bkp_restore_local_create():
+    validate_backup_create("local", namespace, backup_info)
```
Comment (sowmyav27) [inline tests/validation/tests/v3_api/test_bkp_restore_local.py]: we can remove "local" here as it is not specifically handled in the validate_backup_create method

File: tests/validation/tests/v3_api/test_bkp_restore_s3_with_iam.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    namespace["project"] = p
+    namespace["c_client"] = c_client
+
+    def fin():
+        client = get_user_client()
+        cluster_cleanup(client, cluster)
+    request.addfinalizer(fin)
+
+
+@pytest.fixture(scope='module')
```
Comment (sowmyav27) [inline tests/validation/tests/v3_api/test_bkp_restore_s3_with_iam.py]: This need not be marked as a fixture as its being used only once in `create_project_client_and_cluster_s3_with_iam`

**Pull request merged**